lib/src/game/viewport.dart

```diff
@@ -0,0 +1,185 @@
+  /// size of your widget). This is the raw size as it would be without any
+  /// viewport.
+  ///
+  /// You probably don't need to care about this if you are using a viewport.
+  Vector2 getRawSize();
+}
+
+/// This is the default viewport if you want no transformation.
+/// The rawSize is just propagated to the effective size and no
+/// translation is applied.
+/// This basically no-ops the viewport.
+class DefaultViewport extends Viewport {
+  Vector2 _size;
+
+  @override
+  void render(Canvas c, void Function(Canvas c) renderGame) {
+    renderGame(c);
+  }
+
+  @override
+  void resize(Vector2 newRawSize) {
+    _size = newRawSize;
+  }
+
+  @override
+  Vector2 getEffectiveSize() => _size;
+
+  @override
+  Vector2 getRawSize() => _size;
+}
+
+/// This is the most common viewport if you want to have full control of what
+/// the game looks like. Basically this viewport makes sure the ratio between
+/// width and height is *always* the same in your game, no matter the platform.
+///
+/// To accomplish this you choose a virtual size that will always match the
+/// effective size.
+///
+/// Under the hood, the Viewport will try to expand (or contract) the virtual
+/// size so that it fits the most of the screen as it can. So for example,
+/// if the viewport happens to be the same ratio of the screen, it will resize
+/// to fit 100%. But if they are different ratios, it will resize the most it
+/// can and then will add black (color is configurable) borders.
+///
+/// Then, inside your game, you can always assume the game size is the fixed
+/// dimension that you provided.
+///
+/// Normally you can pick a virtual size that has the same ratio as the most
+/// used device for your game (like a pretty standard mobile ratio if you
+/// are doing a mobile game) and then in most cases this will apply no
+/// transformation whatsoever, and if the a device with a different ratio is
+/// used it will try to adapt the best as possible.
+class FixedRatioViewport extends Viewport {
+  Paint borderColor;
+
+  Vector2 rawSize;
+  Vector2 scaledSize;
+  Vector2 resizeOffset;
+  Vector2 fixedSize;
+  double scale;
+
+  FixedRatioViewport(
+    this.fixedSize, {
+    Paint borderColor,
+  }) {
+    this.borderColor = borderColor ?? BasicPalette.black.paint;
+  }
+
+  @override
+  Vector2 getEffectiveSize() => fixedSize;
+
+  @override
+  Vector2 getRawSize() => rawSize;
+
+  @override
+  void resize(Vector2 newRawSize) {
+    rawSize = newRawSize;
+
+    final scaleVector = rawSize.clone()..divide(fixedSize);
+    scale = math.min(scaleVector.x, scaleVector.y);
+
+    scaledSize = fixedSize.clone()..scale(scale);
+    resizeOffset = (rawSize - scaledSize) / 2;
+  }
+
+  @override
+  void render(Canvas c, void Function(Canvas) renderGame) {
+    c.save();
+    c.translate(resizeOffset.x, resizeOffset.y);
+    c.scale(scale, scale);
+
+    renderGame(c);
+
+    c.restore();
+    c.drawRect(
```

Comment:
Suggestion, instead of drawing rects to create the borders, you could use `clip` here, less code and I think it would have better perfomance. We would not even need a border color then, because the background color of the game would already be that.

Reply:
ah good idea, let me learn about clip